### PR TITLE
bash: Eliminate several bugs causing exceptions

### PR DESCRIPTION
**Pull request by mjambon**
Parsing rate just went from 31% to 92%!

Uses https://github.com/returntocorp/pfff/pull/481

PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep-core/src/analyzing/AST_to_IL.ml [hunk 1]
```diff
@@ -891,7 +891,15 @@ let rec stmt_aux env st =
...
       let ss1 = for_var_or_expr_list env xs in
       let st = stmt st_env st in
-      let ss2, cond = expr_with_pre_stmts env (List.nth e 0) (* TODO list *) in
+      let ss2, cond =
+        match e with
+        | first :: _TODO ->
+            (* TODO list *)
+            expr_with_pre_stmts env first
+        | [] ->
+            (* TODO: handle this gracefully *)
```
Comment (aryx) [inline semgrep-core/src/analyzing/AST_to_IL.ml]: you need to return a fixme_todo or something like that, but not raise an exception.

File: semgrep-core/src/cli/Main.ml [hunk 2]
```diff
@@ -1417,10 +1417,14 @@ let main () =
       | [] -> Common.usage usage_msg (options ()))
 
 (*****************************************************************************)
+
+let register_exception_printers () =
```
Comment (aryx) [inline semgrep-core/src/cli/Main.ml:720]: Nice, I didn't know about that. Useful.

File: semgrep-core/src/parsing/ast/Bash_to_generic.ml [hunk 3]
```diff
@@ -299,6 +297,18 @@ and command (env : env) (cmd : command) : stmt_or_expr =
       | arguments ->
           let args = List.map (expression env) arguments in
           Expr (loc, call loc C.cmd args))
+  | And (loc, left, and_tok, right) ->
+      let open_, close = loc in
+      let left = pipeline env left |> as_expr in
+      let right = pipeline env right |> as_expr in
+      let op = G.IdSpecial (G.Op G.And, and_tok) |> G.e in
+      Expr (loc, G.Call (op, (open_, [ G.Arg left; G.Arg right ], close)) |> G.e)
```
Comment (aryx) [inline semgrep-core/src/parsing/ast/Bash_to_generic.ml]: There is a new helper function, AST_generic.opcall that makes this less verbose.

**Review by aryx — APPROVED** [review]

File: semgrep-core/src/parsing/ast/Bash_to_generic.ml (hunk 3 above)
Comment (mjambon) [reply semgrep-core/src/parsing/ast/Bash_to_generic.ml]: nice.

File: semgrep-core/src/analyzing/AST_to_IL.ml (hunk 1 above)
Comment (mjambon) [reply semgrep-core/src/analyzing/AST_to_IL.ml]: ok, I did this.

**Pull request merged**